src/modifierGolemancer.ts

```diff
@@ -0,0 +1,124 @@
+          const newCardId = Unit.unitSourceIdToName(golem_unit_id, false);
+          const upgrade = Upgrade.upgradeCardsSource.find(u => u.title == newCardId)
+          if (upgrade) {
+            underworld.forceUpgrade(player, upgrade, true);
+          } else {
+            console.error('Could not find summon golem upgrade for Golemancer rune');
+          }
+        }
+      });
+    },
+  });
+  registerEvents(golemancerId, {
+    onTurnStart: async (unit: Unit.IUnit, underworld: Underworld, prediction: boolean) => {
+      const modifier = unit.modifiers[golemancerId];
+      if (modifier && unit.alive) {
+        spawnGolems(unit, modifier.quantity, underworld, prediction);
+      }
+    },
+    onSpawn: (unit: Unit.IUnit, underworld: Underworld, prediction: boolean) => {
+      const modifier = unit.modifiers[golemancerId];
+      if (modifier && unit.alive) {
+        spawnGolems(unit, modifier.quantity, underworld, prediction);
+      }
+    }
+  });
+}
+
+function spawnGolems(unit: Unit.IUnit, quantity: number, underworld: Underworld, prediction: boolean) {
+  let allyGolems = getLivingAllyGolems(unit, underworld, prediction);
+  if (allyGolems.length < quantity) {
+    const seed = seedrandom(`${underworld.seed}-${underworld.turn_number}-${unit.id}`);
+    // Summon ally golems up to quantity
+    const golemsToSummon = quantity - allyGolems.length;
+    for (let i = 0; i < golemsToSummon; i++) {
+      const coords = findRandomSummonLocation(unit, unit.attackRange / 2, underworld, prediction, seed)
+      if (coords) {
+        let sourceUnit = allUnits[golem_unit_id];
+        if (sourceUnit) {
+          const summonedUnit = Unit.create(
+            sourceUnit.id,
+            coords.x,
+            coords.y,
+            unit.faction,
+            sourceUnit.info.image,
+            UnitType.AI,
+            sourceUnit.info.subtype,
+            sourceUnit.unitProps,
+            underworld,
+            prediction,
+          );
+
+          makeRisingParticles(summonedUnit, prediction);
+          // Resurrect animation is the die animation played backwards
+          Unit.playAnimation(summonedUnit, summonedUnit.animations.die, { loop: false, animationSpeed: -0.2 });
+          if (summonedUnit.image) {
+            summonedUnit.image.sprite.gotoAndPlay(summonedUnit.image.sprite.totalFrames - 1);
+          }
+        }
+      } else {
+        console.log("Golemancer could not find valid spawn");
+      }
+    }
+  }
+}
+
+export function getLivingAllyGolems(unit: Unit.IUnit, underworld: Underworld, prediction: boolean) {
+  let units = prediction ? underworld.unitsPrediction : underworld.units;
+  units = units.filter(u => u.alive && u.faction == unit.faction && [golem_unit_id, BLOOD_GOLEM_ID].includes(u.unitSourceId));
+  return units;
+}
+
+export function findRandomSummonLocation(unit: Unit.IUnit, radius: number, underworld: Underworld, prediction: boolean, seed: prng): Vec2 | undefined {
```

Comment:
Let's import a similar function rather than making a new one.  If you want something close to the player use underworld.findValidSpawn and if you want something farther import from Summoner or Deathmason

Reply:
I think I had issues with these functions but I didn't want to change them since they were being used in a lot of places already.

I'll look at it again and see what I can do about cleaning up the duplicate code

Reply:
If you have a particular reason for not wanting to use it you don't have to.  I just wasn't sure if you were aware of it.  If you want to go ahead as is that's fine with me. 
Did the issues you have also stand as issues for the summoner? Maybe we should just refactor a single (or few) spawn functions in a util file?)

Reply:
I couldn't use the other functions for Alchemist because I wanted to spawn the potions a specific distance away from the player. Close enough they can grab it, but not so close its immediately picked up if players don't want to pick it up. I also didn't want the placement to be predetermined on a hexgrid because it's an awkward mechanic to play around. I wanted to use a custom radius for Golemancer, and it felt similarly cluttered/awkward with the hex grid imo.

FindValidSpawn seems a bit buggy too, where it sometimes doesn't spawn things? I'm not sure. We got a report about that / last will. There may have been some other logic I wanted to skip from the summoner function, but I haven't looked at it in a couple days now.

I do like the idea of refactoring into one or a couple generic functions in a util file, that would be much cleaner than how I wrote it

Reply:
Okay this works! Thanks for the extra context